### PR TITLE
Resync NTP regurlarly

### PR DESCRIPTION
**Pull request by CurlyMoo**
When the year is 1970 resync every 5 minutes. If the date is properly set, resync every day.

### CODE REVIEW

**Comment by IgorYbema** [comment]
Seems ok

**Comment by CurlyMoo** [comment]
Let's wait at least for the results of the tests of @McMagellan

**Comment by RichieB2B** [comment]
The rescheduling is done twice: both in `timer_cb()` every minute and in `ntpReload()` itself every 5 minutes. So you'll get 2 timerqueue_inserts(), then 4, then 8 etc. Doubling with every call to timer_cb() for the -5 case.

**Comment by CurlyMoo** [comment]
@RichieB2B Two things. The difference in time was indeed wrong. However, the double call to `timerqueue_insert` wasn't necessarily because it would just overwrite the running timer. Despite that, i changed it according to your feedback.

**Comment by RichieB2B** [comment]
@CurlyMoo So an extra call to timerqueue_inserts(x, y, -5) would just reset the previous timer. That's much less of an issue than I thought but still good to avoid it.

**Comment by CurlyMoo** [comment]
@RichieB2B Exactly.

**Comment by McMagellan** [comment]
Feedback: Testing time synchronization. Alpha-467a907, #527
The test failed.

Method:
1) Disconnect Heishamon and Fritzbox from the power supply.
2) Turn the power back on.
3) After about 6 minutes, contact with the Heishamon console could be established again from the browser. 
--> Timestamp 1/1/1970, 1:06, Uptime 6 minutes.
5) Today morning see Screenshot: 
-> Timestamp 1/1/1970, 9:37, Uptime 8:h 37 min.

Screenshot 2024-05-19 at 07-40-51 Heisha monitor

**Comment by geduxas** [comment]
@CurlyMoo why not use some basic logic, just 2 variables, one boolen for time sync'ed flag, and one last time synced (timestamp after timesync) and check by that.. if it's never synced (try every 5min) and if synced try after last sync_time timeout.

**Comment by CurlyMoo** [comment]
> @CurlyMoo why not use some basic logic, just 2 variables, one boolen for time sync'ed flag, and one last time synced (timestamp after timesync) and check by that.. if it's never synced (try every 5min) and if synced try after last sync_time timeout.

That's what we're doing here, but centralized through the timepool.

**Comment by geduxas** [comment]
Oh, sorry, i saw somewhere you mentioned exact dates.. so thought it tied to dates :) didn't check source.

**Comment by CurlyMoo** [comment]
> Oh, sorry, i saw somewhere you mentioned exact dates.. so thought it tied to dates :) didn't check source.

It does, because by definition unix timestamp start on the 1st of januari 1970. So, checking for a year other than 1970 is similar to that sync flag but with more accuracy of success.

https://en.wikipedia.org/wiki/Unix_time

**Comment by CurlyMoo** [comment]
@McMagellan Can you let it run again for a while and check for the `Syncing with ntp servers` loglines?

**Comment by McMagellan** [comment]
Feedback: Testing time synchronization. Alpha baddb94, #531
The test failed again.

How I do it:
1) Close the console window in the PC browser.
2) Disconnect the main Fritzbox1 from power
3) (in the house basement) switch off the power to the Fritzbox2 repeater.
4) Switch off Heishamon, wait 1 minute.
5) Switch on Heishamon, switch on Fritzbox2, switch on Fritzbox1.

LED WLAN and DSL on Fritzbox1 flash for about 5 - 6 minutes until the PC browser is connected to the WLAN again.
During this time the Heishamon Setup AP is active and I connected to the console via it to test it.
Message every 15 seconds: Reconnecting to WIFI failed. Waiting a few seconds before trying again.

Screenshot 2024-05-19 at 11-24-56 Heisha monitor


For the next attempt, I activated "Debug log to MQTT topic from start" to have the messages recorded in the influxdb immediately after reconnection.
In the log file (see logfile) the first 4 lines are from the previous correct connection. The first data set is recorded after 3 minutes 11 seconds after power returns (start uptime new). 
[panasonic heat pump_log.csv](https://github.com/IgorYbema/HeishaMon/files/15369470/panasonic.heat.pump_log.csv)

The timestamp at the beginning of the line comes from the Raspberry. 
The Heishamon time is then at 1/1/1970 and 1:03:11 and runs synchronously with the uptime. 
There was no further synchronization.

**Comment by McMagellan** [comment]
@CurlyMoo: How can i do this?
@McMagellan Can you let it run again for a while and check for the Syncing with ntp servers loglines?

**Comment by stumbaumr** [comment]
Somehow yesterday afternoon the HeishaMon rebooted, left rules active, but did not get the proper time. Only found out this morning when looking at the graphs ans seeing a DHW production attempt at 4:55. Looks like I need to help to get this PR into main...
grafik

**Comment by CurlyMoo** [comment]
> Looks like I need to help to get this PR into main...

Please do!

**Comment by stumbaumr** [comment]
Why do you need two timers? The code from timer -6 could look like this:
```
      case -6: {
          time_t now = time(NULL);
          struct tm *tm_struct = localtime(&now);
          if(tm_struct->tm_year == 1970) {
            logprintln_P(F("I am still in the year 1970, syncing with ntp servers, check again in one minute"));
            ntpReload(&heishamonSettings);
            timerqueue_insert(60, 0, -6);
          } else {
            logprintln_P(F("Time probably correct, syncing with ntp servers, check again in a day"));
            ntpReload(&heishamonSettings);
            timerqueue_insert(86400, 0, -6);
          }
        } break;
```
and then in systemboot just call it after a minute.
```
      logprintln_P(F("Syncing with ntp servers, check again in one minute"));
      ntpReload(&heishamonSettings);
      timerqueue_insert(60, 0, -6);
```

And then delete timer -5...
...and rename -6 to -5...

**Comment by CurlyMoo** [comment]
> Why do you need two timers?

Because in your case you're always doing a `ntpReload` even if it's not necessary.

**Comment by stumbaumr** [comment]
> > Why do you need two timers?
> 
> Because in your case you're always doing a `ntpReload` even if it's not necessary.

Only once, right after the systemboot.
But you could also remove that, doing things like this
```
      case -6: {
          ntpReload(&heishamonSettings);
          time_t now = time(NULL);
          struct tm *tm_struct = localtime(&now);
          if(tm_struct->tm_year == 1970) {
            logprintln_P(F("I am still in the year 1970, check again in one minute"));
            timerqueue_insert(60, 0, -6);
          } else {
            logprintln_P(F("Time probably correct, check again in a day"));
            timerqueue_insert(86400, 0, -6);
          }
        } break;
```
and
```
      logprintln_P(F("Syncing with ntp servers after boot"));
      timerqueue_insert(1, 0, -6);
```

**Comment by CurlyMoo** [comment]
Same thing. You are already doing a ntpReload while you haven't checked if the previous ntpReload succeeded.

**Comment by CurlyMoo** [comment]
And i don't see why using two timers is an issue? They're not active at the same time so aren't a performance issue at all.

**Comment by stumbaumr** [comment]
> Same thing. You are already doing a ntpReload while you haven't checked if the previous ntpReload succeeded.

It is just doing it the other way around:
- Do the ntpReload, if it failed, schedule another one and try then again.
- If successful, schedule it for the next day anyway.

> And i don't see why using two timers is an issue? They're not active at the same time so aren't a performance issue at all.

It is less code and easier to read. Less code is easier to maintain.

**Comment by CurlyMoo** [comment]
No it isn’t, because ntpReload is an async function.

**Comment by stumbaumr** [comment]
Oh, ok... sorry. My spaghetti-code-brain has problems processing things like that and I didn't know it is async...

**Comment by IgorYbema** [comment]
Please confirm this one is working and ok to be merged into release v3.7

**Comment by McMagellan** [comment]
Feedback Test Igor #580 Signature Alpha3c3b16e - Time Sync.

Time synchronization after power off failed again.

I performed the following test:
1) Fritzbox1 with DSL connection power off.
2) Fritzbox2 (Mesh Repeater) power off.
3) Heishamon V4 power off.

4) Heishamon power up
5) Fritzbox2 power up
6) Fritzbox1 (DSL) power up.

The next 30 minutes are recorded in the following debug log file. Start at HeishamonPower up.
Manual reboot after 26 minutes with incorrect time. -> Correct start.

[25cputty.log](https://github.com/user-attachments/files/16741626/25cputty.log)

543: 
Thu Jan 1 01:00:14 1970 (15002): Starting initial wifi scan...
Thu Jan 1 01:00:15 1970 (15671): WiFi lost, starting setup hotspot...
Thu Jan 1 01:00:15 1970 (15780): Reconnecting to WiFi failed. Waiting a few seconds before trying again.

652:
Thu Jan 1 01:00:45 1970 (45673): Reconnecting to WiFi failed. Waiting a few seconds before trying again.
(every 30 seconds until 821:)

821:
Thu Jan 1 01:02:30 1970 (150752): Retrying configured WiFi, ...

827: 
Thu Jan 1 01:02:35 1970 (155038): WiFi (re)connected, shutting down hotspot...

After 2 minutes 35 seconds, Heishamon connects to the Fritzbox2 (Mesh Repeater). At this point, the Fritzbox1 as a WiFi master does not yet have internet. (DSL not ready)

**Log from Fritzbox1 (DSL trainig)**
DSL training starts at 20:03:36.
DSL connected at 20:06:55.

Screenshot 2024-08-25 at 20-18-35 FRITZ!Box 7490

**Log from Friztbox2 (Mesh Repeater)**
Heishamon is the second device who connects to the network (20:04:21)
Minutes later is the time synchronisation done. (20:06:56)

Screenshot 2024-08-25 at 20-25-01 FRITZ!Box 7490

**Comment by IgorYbema** [comment]
This fix isn't in #580 and the build file is not available anymore. Need to create a new build file

**Comment by McMagellan** [comment]
Feedback Test Igor #580 Signature Alpha3c3b16e 
I'm not sure if this is the right place for this error message and firmware #580.
I think this is about the storage problem between rules and the web server.

I simply let the #580 continue to run after yesterday's test and it crashed.
See debug log file. Numbers are line numbers in the debug file.
[26a1putty.txt](https://github.com/user-attachments/files/16749678/26a1putty.txt)


With an uptime of 16 hours 41 minutes, the message occurs in the middle of the TOP bulk phase according to TOP86 and without a WS connection having been established beforehand.
924: Out of memory webserver_send_content:#1688 --> Crash#1

Heishamon then automatically starts correctly and loads the ruleset correctly.

I then opened Heishamon 10 minutes later and the homepage froze, which is why I only noticed the problem.
2425: New webserver client: 192.168.178.86:59878
2429: Heishamon Crash#2

Heishamon then restarts automatically **but without loading the rules**.
2498: Not loading rules due to crash reboot!

In order to restore normal operation, I then rebooted via the browser.
2842: Starting debugging, version: Alpha-3c3b16e
2898: 2149: reading rules
3063: 4486: ==== System#Boot ====

**Comment by IgorYbema** [comment]
> Feedback Test Igor #580 Signature Alpha3c3b16e I'm not sure if this is the right place for this error message and firmware #580. I think this is about the storage problem between rules and the web server.

Is not the right place as we are discussing the NTP release here. But I did check and you had a webclient connected just before the first crash. (about 13:08: New webserver client: 192.168.178.86:59842). 
Just wait for the websocket update to fix the memory issue's.

Not sure why it crashed again at 13:18 but that also came after a webserver client connection. The webserver code is still not that bug free I am afraid.

But lets discuss NTP updates from now on in this PR

**Comment by McMagellan** [comment]
Feedback Test Igor #583 Signature Alphaf2a171b (NTP Sync)

I repeated yesterday's test twice with firmware #583.
Here is the debug log file from the second test.
[26dputty.log](https://github.com/user-attachments/files/16752065/26dputty.log)


The time has been synchronized.
During this test I deliberately avoided any web server calls.
No rules were loaded either.

What is interesting in the screenshot of the Fritzbox is that Heishamon (192.168.178.36) only connected to the WLAN at 18:55:34 even though it had been available since 18:40:24.

Screenshot 2024-08-26 at 19-28-16 FRITZ!Box 7490


The first test went so badly that time synchronization didn't even happen.
When rules were loaded, a crash occurred immediately after every web server call.
In the Logfile you can see the chaos when rules are loaded abd WS connected.
[26cputty.log](https://github.com/user-attachments/files/16752067/26cputty.log)

**Comment by McMagellan** [comment]
@IgorYbema > Together with a big ruleset that can cause a memory issue which I believe is the case.

How big can a ruleset be or how big can an individual rule be? I am using Heishamon V4.
Your documentation says: 
`The techniques used in the rule library allows you to work with very large rulesets, but best practice is to keep it below 10,000 bytes.`

When I load my ruleset, the response appears in the console: 134793: rules memory used: 5892 / 16384
My ruleset is now 415 lines long and I need about 500 lines to finish it.

Will it be possible to implement this or should I switch to Blockly?

**Comment by IgorYbema** [comment]
Not sure yet what the limit is but lets wait for the websocket update to be finished so the webcalls take less memory.
Anyway, if you are not using the web interface it shouldn't be a problem anyway (only if your automation uses the /json calls instead of mqtt). The webinterface normally only used to commission the system and to enter the rules. After that you don't need to have it running

**Comment by stumbaumr** [comment]
> I added this PR commits to v3.7 branch just now so the resulting binary #583 can be tested.

Installed it just now and will commence NTP tests on Wednesday - as soon as most family members are away...

**Comment by stumbaumr** [comment]
First test unsuccessful:
- Powered off/on the heatpump, switch and Fritzbox using the fuses in the basement
- HeishaMon started up and had for over 5 minutes 1970 as year
- Then at some point HeishaMon rebooted and disabled the rules, but since Fritzbox was already online it now got the correct date.

I had the console open the whole time. Trying again without the web ui as soon as I finish this glass of wine..

**Comment by stumbaumr** [comment]
13 Minutes after power off/on I had a look at the console and it still showed 1970 as the year.

And then it rebooted and deactivated the rules again...

**Comment by McMagellan** [comment]
The same as mine. You are not allowed to establish a connection to the web server, rules are not allowed to be loaded from previous firmware, nor are you allowed to reload rules using the Save button.
In order to see what happens, you need the output from the debug port, ideally saved in a log file. If you wait about 15 minutes you will see that it worked.

Take a look at the log file 26cputty.log from yesterday, in which the whole chaos was recorded, which you also have with you. The NTP change doesn't have any effect here because Heishamon crashes beforehand and then reboots at some point when the time sync is available.

But what exactly happened is recorded in the debug log file 26dputty.log.
The problem here is still that the web server environment and the rules environment are fighting for memory.
The problems are known and will be fixed.

**Comment by stumbaumr** [comment]
I tried to prove your theory to let it run longer without using the web interface but even after well over fourty minutes the ntp is not correcting the time...

See MQTT log output:
image

**Comment by stumbaumr** [comment]
A reboot while Fritzbox being up just works...

image

File: HeishaMon/HeishaMon.ino [hunk 1]
```diff
@@ -1046,6 +1048,26 @@ void timer_cb(int nr) {
...
+          struct tm *tm_struct = localtime(&now);
+          if(tm_struct->tm_year == 1970) {
+            ntpReload(&heishamonSettings);
+            logprintln_P(F("Syncing with ntp servers, check again in 5 minutes"));
+            timerqueue_insert(300, 0, -5);
+          } else {
+            /*
+             * Wait 300 sec less than a full day
+             */
+            logprintln_P(F("Syncing with ntp servers, check again in a day"));
```
Comment (CurlyMoo) [inline HeishaMon/HeishaMon.ino:1067]: This log should change to:
`Succesfully synced with ntp servers, checking again in a day`

**Comment by CurlyMoo** [comment]
In the logs by @McMagellan it works fine everytime. @IgorYbema The only thing that confuses me is that in the `26dputty.log` it started its first sync after 17 minutes. However, @McMagellan could connect to the webserver in the meanwhile. So @IgorYbema either i put the first sync in the wrong spot or in the `check_wifi` function it doesn't enter this line right: https://github.com/IgorYbema/HeishaMon/blob/efec85685d01dfd2de115ed0d7809b1d2bb3a2ce/HeishaMon/HeishaMon.ino#L172

**Comment by McMagellan** [comment]
What I also saw is a hole in the MQTT bulk transmission shortly after the correct time synchronization.

Thu Aug 29 13:15:59 2024 (11160): received TOP104 Solar_Frost_Protection: 5
Thu Aug 29 13:15:59 2024 (11165): received TOP105 Solar_High_Limit: 80
Thu Aug 29 13:15:59 2024 (11173): received TOP106 Pump_Flowrate_Mode: 1
Thu Aug 29 13:15:59 2024 (11178): received TOP107 Liquid_Type: 0
Thu Aug 29 13:15:59 2024 (11185): received TOP108 Alt_External_Sensor: 0
Thu Aug 29 13:15:59 2024 (11190): received TOP109 Anti_Freeze_Mode: 0
Thu Aug 29 13:15:59 2024 (11197): received TOP110 Optional_PCB: 0
Thu Augeived TOP115 Water_Pressure: -0.02
Thu Aug 29 13:15:59 2024 (11234): received TOP116 Second_Inlet_Temp: -128
Thu Aug 29 13:15:59 2024 (11240): received TOP117 Economizer_Outlet_Temp: -128
Thu Aug 29 13:15:59 2024 (11248): received TOP118 Second_Room_Thermostat_Temp: -128
Thu Aug 29 13:16:03 2024 (15002): Starting initial wifi scan ...
Thu Aug 29 13:16:08 2024 (20002): Requesting new 1wire temperatures

TOP111 to TOP115 are missing here. Never seen bevore.

**Comment by CurlyMoo** [comment]
Let's some out a bit. Until now we have been suspecting the NTP additions are the ones that are faulty, but i would like to focus first on correct functioning of the `check_wifi` function.

@McMagellan @stumbaumr can you guys monitor in the difference tests how long it takes for the wifi to connect, how long it takes for the first ntp sync message to appear in the log, and how long it takes before the time is set correctly?

**Comment by CurlyMoo** [comment]
You can clearly see it in the last log `29bputty.log`. Until the reboot/crash there is just no wifi connection. Wifi stays at `-1` in the stats. As soon as it get a wifi connection after the reboot the time is synced immediatly and the wifi is at `100%`.

So focussing on the ntp sync without a proper wifi connection is useless.

**Comment by McMagellan** [comment]
Here is the information from the system log of the two Fritz boxes after power on at 12:35 (time stamp from the start of the Putty file). Sorry it's German.

[Systemlog Fritzbox1 DSL.pdf](https://github.com/user-attachments/files/16800659/Systemlog.Fritzbox1.DSL.pdf)
[Systemlog Fritzbox2 Mesh.pdf](https://github.com/user-attachments/files/16800662/Systemlog.Fritzbox2.Mesh.pdf)

The first message in the log is around 12:37, although it is unclear to me where the Fritzbox got this time because no ntp sync could take place yet. Maybe from the DSL Training Start Message.

The first WiFi device logs on to the network at 12:37:08.

At 12:40:10 DSL IP V4 was connected to Fritzbox1 

At 12:40:11 the message about successful time synchronization comes after two unsuccessful attempts.

Heishamon only logged on to the network at 13:15:56 after the manual reboot with the IP address 192.168.178.36.

**Comment by IgorYbema** [comment]
The reason why it didn't connect to wifi is, I believe, due to that you have connected to the AP:
```
Thu Jan  1 01:00:37 1970 (37546): WiFi lost, but softAP station connecting, so stop trying to connect to configured ssid...
New webserver client: 192.168.4.100:59214
```
However, it should not report this message after that:
`Thu Jan  1 01:00:45 1970 (45637): Reconnecting to WiFi failed. Waiting a few seconds before trying again.
`
I think that last one is just a bug. I'll check that. But remember that it will not retry to connect to the wifi if a client is connected (192.168.4.*) to the setup AP.

I'll check the NTP code

**Comment by IgorYbema** [comment]
> In the logs by @McMagellan it works fine everytime. @IgorYbema The only thing that confuses me is that in the `26dputty.log` it started its first sync after 17 minutes. However, @McMagellan could connect to the webserver in the meanwhile. So @IgorYbema either i put the first sync in the wrong spot or in the `check_wifi` function it doesn't enter this line right:
> 
> https://github.com/IgorYbema/HeishaMon/blob/efec85685d01dfd2de115ed0d7809b1d2bb3a2ce/HeishaMon/HeishaMon.ino#L172

```
This is all correct. It started syncing just after the first wifi reconnect (after 17 mins):
Thu Jan  1 01:17:19 1970 (1039076): WiFi (re)connected, shutting down hotspot...
1039187: Syncing with ntp servers, check again in 5 minutes
Mon Aug 26 18:55:33 2024 (1040090): Requesting new 1wire temperatures
```

**Comment by IgorYbema** [comment]
Seems you do have a client connecting to it. Maybe a phone you used before and is now auto connecting?

**Comment by McMagellan** [comment]
In the past, I actually used the Heishamon AP to connect my cell phone to the normal WiFi in the settings in the event of a factory reset. I will repeat the test and first turn off my cell phone.

**Comment by McMagellan** [comment]
Feedback new Test: ntp sync.

The test was manually stopped after 20 minutes without the time being synchronized.
I turned off my phone to make sure no web server calls were coming from it.
Then I did an interim test and saw WS calls to the AP in the log again. (New webserver client: 192.168.4.100:59457)
I hadn't considered the old Lenovo on which the debug putty files are recorded and which is in the immediate vicinity. 

However, all network connections were then disconnected. 
Nevertheless, he made calls to the AP. 
I then deactivated the WiFi adapter in the Lenovo and started the test again.

There is only one webserver call in the 20 minutes:
Thu Jan 1 01:08:31 1970 (511410): Checksum and header received ok! ` (only timestamp info)`
514017: Syncing with ntp servers, check again in a day
New web server client: 192.168.178.97:43937
Closing web server client: 192.168.178.97:43937
Thu Jan 1 01:08:40 1970 (520051): Requesting new 1wire temperatures

This IP address is the Fritzbox2 which runs as a MESH- repeater.
For more details see the log file.
[29cputty2.log](https://github.com/user-attachments/files/16805358/29cputty2.log)

**Comment by IgorYbema** [comment]
It reconnected ok. The time sync failed. Was DSL trained already?

`hu Jan  1 01:03:33 1970 (214019): WiFi (re)connected, shutting down hotspot...
214130: Syncing with ntp servers, check again in 5 minutes`

**Comment by CurlyMoo** [comment]
What i also find wierd is that the subsequent check is suddenly:
`514017: Syncing with ntp servers, check again in a day`

Whilst the year is still 1970.

**Comment by McMagellan** [comment]
@CurlyMoo: The timing in the Fritzbox is always the same. There are only minimal time differences between the network master and the repeater (MESH).
0:00 Power up
2:30 Fritzbox starts service
2:30 Training DSL start (first timestamp in Fritzbox log)
3:30 WiFi starts
5:45 DSL connected Internet
5:46 Time server received.

To me it looks like Heishamon checks whether the last network connection is available just a few seconds after starting. 
If this is not the case, the AP is activated. 
If devices now find the AP even without actually connecting, no attempt will be made to connect to the last network. Heishamon should determine that these connections are not real connections and then continue trying to reestablish the last connection. For me this would be possible around 3:30 after powering up the Fritzbox after the WLAN went into operation. However, the correct time is only available after 5:46.

**Comment by IgorYbema** [comment]
So Heishamon reconnected at 3:33 to the wifi. No not at that time. Retry at 8:33 failed to sync. Thanks. I'll try to find out why

**Comment by IgorYbema** [comment]
> It reconnected ok. The time sync failed. Was DSL trained already?
> 
> `hu Jan 1 01:03:33 1970 (214019): WiFi (re)connected, shutting down hotspot... 214130: Syncing with ntp servers, check again in 5 minutes`

So probably the year == 1970 check is wrong. That is why it flips over to next day check and doesnt run a ntpReload.

**Comment by IgorYbema** [comment]
> > What i also find wierd is that the subsequent check is suddenly:
> > 514017: Syncing with ntp servers, check again in a day
> > Whilst the year is still 1970.
> 
> So probably the year == 1970 check is wrong. That is why it flips over to next day check and doesnt run a ntpReload.

tm_year = years since 1900... 

Patching this stupid one

**Comment by CurlyMoo** [comment]
It again proves the need of reviews.

File: HeishaMon/HeishaMon.ino [hunk 2]
```diff
@@ -1050,21 +1050,24 @@ void timer_cb(int nr) {
         } break;
       case -5: {
           ntpReload(&heishamonSettings);
-          logprintln_P(F("Syncing with ntp servers, check again in 5 minutes"));
-          timerqueue_insert(300, 0, -6);
+          logprintln_P(F("Resynced with NTP servers. Next sync after 24 hours."));
+          timerqueue_insert(86400, 0, -5);
```
Comment (CurlyMoo) [inline HeishaMon/HeishaMon.ino:1054]: With this change, you're 100% sure that the time cannot become bogus again and reset to it's defaults.

**Comment by McMagellan** [comment]
Feedback #587 ntp sync. Signature Alpha b24fab5

I ran 2 tests of 20 minutes each and in both cases the time was synchronized correctly after about 7 minutes.
The second test was with a cell phone and the WiFi adapter was active again on the Lenovo. There were no more web server connections.
[30bputty.log](https://github.com/user-attachments/files/16821102/30bputty.log)
[30cputty.log](https://github.com/user-attachments/files/16821104/30cputty.log)

Here too, when I was looking through the log file 30bputty.log, I noticed a hole in the MQTT bulk message.
Line 179: TOP110 to TOP115 are missing.
Line 451: TOP109 to TOP113 are missing
We've had this happen in the last few days.

File: HeishaMon/HeishaMon.ino (hunk 2 above)
Comment (IgorYbema) [reply HeishaMon/HeishaMon.ino:1054]: I see no reason outside a reboot that the time will get bogus indeed

Comment (CurlyMoo) [reply HeishaMon/HeishaMon.ino:1054]: Let's see if this fixes this unnecessary sync from the logs from @McMagellan 
```
Fri Aug 30 19:53:24 2024 (751365): Received 203 bytes data
Fri Aug 30 19:53:24 2024 (751366): Checksum and header received ok!
Fri Aug 30 19:53:24 2024 (751368): received TOP49 Main_Hex_Outlet_Temp: 23
754921: Syncing with ntp servers, check again in 5 minutes
```

Comment (stumbaumr) [reply HeishaMon/HeishaMon.ino:1054]: Lovely work!!!

After boot:
image

And then...
image

I believe the PR is done!!!

**Comment by IgorYbema** [comment]
> Here too, when I was looking through the log file 30bputty.log, I noticed a hole in the MQTT bulk message. Line 179: TOP110 to TOP115 are missing. Line 451: TOP109 to TOP113 are missing We've had this happen in the last few days.

This is probably just caused by some overloading of the log lines, busy mqtt server or network drops. Nothing to worry about

**Comment by stumbaumr** [comment]
Lovely work!!!

After boot:
image

And then:
image

I believe this PR is done now.

**Comment by IgorYbema** [comment]
will be added to v3.7

**Pull request closed without merging**